base/rational.jl

```diff
@@ -50,20 +74,38 @@ julia> (3 // 5) // (2 // 1)
 //(n::Integer,  d::Integer) = Rational(n,d)
 
 function //(x::Rational, y::Integer)
-    xn,yn = divgcd(x.num,y)
-    xn//checked_mul(x.den,yn)
+    xn, yn = divgcd(x.num,y)
+    num, den = promote(xn, checked_mul(x.den, yn))
+    if signbit(den)
+        den = -den
+        signbit(den) && __throw_rational_argerror_typemin(typeof(den))
+        num = -num
+    end
+    Rational(num, den; safe=false)
 end
 function //(x::Integer,  y::Rational)
-    xn,yn = divgcd(x,y.num)
-    checked_mul(xn,y.den)//yn
+    xn, yn = divgcd(x,y.num)
+    num, den = promote(checked_mul(xn, y.den), yn)
+    if signbit(den)
+        den = -den
+        signbit(den) && __throw_rational_argerror_typemin(typeof(den))
+        num = -num
+    end
+    Rational(num, den; safe=false)
 end
 function //(x::Rational, y::Rational)
     xn,yn = divgcd(x.num,y.num)
     xd,yd = divgcd(x.den,y.den)
-    checked_mul(xn,yd)//checked_mul(xd,yn)
+    num, den = promote(checked_mul(xn, yd), checked_mul(xd, yn))
+    if signbit(den)
```

Comment:
It's not ideal to need to repeat this code. My thinking was that checking the sign bit is so cheap we might as well do it every time in the constructor, regardless of the `safe` setting.

Reply:
True, that's just bad programming. Either your option or I simply add a
```julia
function signcheck(num, den)
    if signbit(den)
        den = -den
        signbit(den) && __throw_rational_argerror_typemin(typeof(den))
        num = -num
    end
    Rational(num, den; safe=false)
end
```
and call it in the four cases where it's needed. Let me know your preference and I'll update accordingly.

Reply:
The code repetition is fixed in the last commit.